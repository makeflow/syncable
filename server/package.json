{
  "name": "@syncable/server",
  "version": "2.0.0-wip.40",
  "publishConfig": {
    "access": "public"
  },
  "description": "Syncable.",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/makeflow/syncable.git"
  },
  "keywords": [
    "syncable"
  ],
  "author": "Chengdu Mufan Technology Co., Ltd.",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/makeflow/syncable/issues"
  },
  "homepage": "https://github.com/makeflow/syncable#readme",
  "main": "bld/library/index.js",
  "types": "bld/library/index.d.ts",
  "files": [
    "src/**/*.ts",
    "bld"
  ],
  "peerDependencies": {
    "@syncable/core": "^2.0.0-wip.0",
    "socket.io": "2"
  },
  "dependencies": {
    "lodash": "^4.17.11",
    "rxjs": "^6.3.3",
    "tslang": "^0.1.11",
    "uuid": "^3.3.2",
    "villa": "^0.3.1"
  },
  "devDependencies": {
    "@types/lodash": "^4.14.119",
    "@types/socket.io": "^1.4.36",
    "@types/uuid": "^3.4.3",
    "socket.io": "^2.1.1"
  }
}
