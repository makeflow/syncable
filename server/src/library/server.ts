import {EventEmitter} from 'events';

import {
  ChangePacket,
  ChangePacketId,
  ChangePlant,
  ChangePlantBlueprint,
  ChangePlantProcessingResultWithTimestamp,
  Context,
  GeneralChange,
  GeneralSyncableRef,
  IChange,
  INotification,
  IRPCDefinition,
  ISyncable,
  ISyncableObject,
  ISyncableObjectProvider,
  IUserSyncableObject,
  SyncableManager,
  SyncableRef,
} from '@syncable/core';
import _ from 'lodash';
import {Server as SocketServer} from 'socket.io';
import uuid from 'uuid';
import * as v from 'villa';

import {Connection, ConnectionSocket} from './connection';

export type RPCFunction<
  TContext extends Context,
  TRPCDefinition extends IRPCDefinition
> = (
  context: TContext,
  manager: SyncableManager,
  params: TRPCDefinition['params'],
) => TRPCDefinition['return'];

export type RPCObject<
  TContext extends Context = Context,
  TRPCDefinition extends IRPCDefinition = IRPCDefinition
> = {
  [K in TRPCDefinition['name']]: RPCFunction<
    TContext,
    Extract<TRPCDefinition, {name: K}>
  >
};

export type ViewQueryFilter<T extends ISyncableObject = ISyncableObject> = (
  object: T,
) => boolean;

export interface ConnectionSession<TViewQuery> {
  group: string;
  userRef: SyncableRef<IUserSyncableObject>;
  viewQuery: TViewQuery | undefined;
}

export interface IGroupClock {
  next(): Promise<number>;
}

interface GroupInfo<TServerGenerics extends ServerGenericParams> {
  clock: IGroupClock;
  manager: SyncableManager;
  connectionSet: Set<Connection<TServerGenerics>>;
  loadingPromise: Promise<void>;
}

export interface ServerGenericParams {
  user: IUserSyncableObject;
  syncableObject: ISyncableObject;
  change: IChange;
  viewQuery: unknown;
  rpcDefinition: IRPCDefinition;
  notification: INotification;
}

interface Server<TGenericParams extends ServerGenericParams> {
  on(event: 'error', listener: (error: Error) => void): this;
  on(
    event: 'notify',
    listener: (
      notification: TGenericParams['notification'],
      id: ChangePacketId,
    ) => void,
  ): this;

  emit(event: 'error', error: Error): boolean;
  emit(
    event: 'notify',
    notification: TGenericParams['notification'],
    id: ChangePacketId,
  ): boolean;
}

export interface ServerOptions<TSyncable extends ISyncable> {
  builtInSyncables?: TSyncable[];
}

abstract class Server<
  TGenericParams extends ServerGenericParams
> extends EventEmitter {
  private server: SocketServer;
  private groupInfoMap = new Map<string, GroupInfo<TGenericParams>>();

  private context = new Context<TGenericParams['user']>('server', 'server');

  private changePlant: ChangePlant;

  constructor(
    server: SocketServer,
    private provider: ISyncableObjectProvider,
    blueprint: ChangePlantBlueprint<TGenericParams>,
    readonly rpc: RPCObject<
      Context<TGenericParams['user']>,
      TGenericParams['rpcDefinition']
    >,
    private options: ServerOptions<
      TGenericParams['syncableObject']['syncable']
    > = {},
  ) {
    super();

    this.server = server;

    this.changePlant = new ChangePlant(blueprint, provider);

    this.initialize().catch(this.errorHandler);
  }

  abstract getViewQueryFilter(
    query: TGenericParams['viewQuery'],
    context: Context,
    manager: SyncableManager,
  ): ViewQueryFilter<TGenericParams['syncableObject']>;

  async update(
    group: string,
    change: TGenericParams['change'],
  ): Promise<ChangePlantProcessingResultWithTimestamp> {
    await this.initializeGroup(group);

    let packet: ChangePacket = {
      id: uuid() as ChangePacketId,
      createdAt: Date.now(),
      ...(change as GeneralChange),
    };

    return this._applyChangePacket(group, packet, this.context);
  }

  applyChangePacket(
    group: string,
    packet: ChangePacket,
    context: Context<TGenericParams['user']>,
  ): void {
    this._applyChangePacket(group, packet, context).catch(this.errorHandler);
  }

  getNextTimestamp(group: string): Promise<number> {
    let {clock} = this.groupInfoMap.get(group)!;
    return clock.next();
  }

  protected abstract createGroupClock(group: string): IGroupClock;

  protected abstract resolveSession(
    socket: ConnectionSocket,
  ): Promise<ConnectionSession<TGenericParams['viewQuery']>>;

  protected abstract loadSyncables(group: string): Promise<ISyncable[]>;

  protected abstract saveSyncables(
    updates: ISyncable[],
    creations: ISyncable[],
    removals: SyncableRef[],
  ): Promise<void>;

  protected errorHandler = (error: Error): void => {
    console.error(error);
    this.emit('error', error);
  };

  private async initialize(): Promise<void> {
    this.server.on('connection', (socket: ConnectionSocket) => {
      this.initializeConnection(socket).catch(console.error);
    });
  }

  private async initializeConnection(socket: ConnectionSocket): Promise<void> {
    let {group, userRef, viewQuery} = await this.resolveSession(socket);

    let {manager, connectionSet} = await this.initializeGroup(group);

    let connection = new Connection<TGenericParams>(
      group,
      socket,
      this,
      manager,
    );

    connectionSet.add(connection);

    socket.on('disconnect', cleanUp);
    socket.on('error', cleanUp);

    connection.initialize(userRef, viewQuery).catch(console.error);

    function cleanUp(): void {
      connectionSet.delete(connection);
    }
  }

  private async initializeGroup(
    group: string,
  ): Promise<GroupInfo<TGenericParams>> {
    let groupInfoMap = this.groupInfoMap;

    let groupInfo = groupInfoMap.get(group);

    if (!groupInfo) {
      let clock = this.createGroupClock(group);
      let manager = new SyncableManager(this.provider);
      let {builtInSyncables = []} = this.options;

      for (let builtInSyncable of builtInSyncables) {
        manager.addSyncable(builtInSyncable);
      }

      let loadingPromise = this.loadAndAddSyncables(group, manager);

      groupInfo = {
        clock,
        manager,
        connectionSet: new Set(),
        loadingPromise,
      };

      groupInfoMap.set(group, groupInfo);
    }

    await groupInfo.loadingPromise;

    return groupInfo;
  }

  private async loadAndAddSyncables(
    group: string,
    manager: SyncableManager,
  ): Promise<void> {
    let syncables = await this.loadSyncables(group);

    for (let syncable of syncables) {
      manager.addSyncable(syncable);
    }
  }

  private async _applyChangePacket(
    group: string,
    packet: ChangePacket,
    context: Context<TGenericParams['user']>,
  ): Promise<ChangePlantProcessingResultWithTimestamp> {
    let info = this.groupInfoMap.get(group);

    if (!info) {
      throw new Error(
        `Syncable group "${info}" has not been initialized on this instance`,
      );
    }

    let {manager, clock} = info;

    let timestamp = await clock.next();

    let {id, refs: refDict} = packet;

    let syncableObjectOrCreationRefDict = _.mapValues(
      refDict,
      (ref: GeneralSyncableRef) =>
        ref
          ? 'creation' in ref && ref.creation
            ? ref
            : manager.requireSyncableObject(ref)
          : undefined,
    );

    let result = this.changePlant.process(
      packet,
      syncableObjectOrCreationRefDict,
      context,
      manager,
      timestamp,
    );

    let {updates, creations, removals, notifications} = result;

    for (let {snapshot} of updates) {
      manager.updateSyncable(snapshot);
    }

    for (let syncable of creations) {
      manager.addSyncable(syncable);
    }

    for (let ref of removals) {
      manager.removeSyncable(ref);
    }

    await this.saveAndBroadcastChangeResult(group, result);

    for (let notification of notifications) {
      this.emit('notify', notification, id);
    }

    return result;
  }

  private async saveAndBroadcastChangeResult(
    group: string,
    result: ChangePlantProcessingResultWithTimestamp,
  ): Promise<void> {
    return v.lock(group, async () => {
      let {updates: updateDict, creations, removals} = result;

      let updates = Object.values(updateDict).map(update => update.snapshot);

      await this.saveSyncables(updates, creations, removals);

      let {connectionSet} = this.groupInfoMap.get(group)!;

      for (let connection of connectionSet) {
        connection.handleChangeResult(result);
      }
    });
  }
}

export interface IServer<TGenericParams extends ServerGenericParams>
  extends Server<TGenericParams> {}

export const AbstractServer = Server;
