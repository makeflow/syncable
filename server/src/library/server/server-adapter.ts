import {
  ChangePacketId,
  ChangePlantProcessingResultUpdateItem,
  ISyncable,
  SyncableRef,
} from '@syncable/core';
import {Observable} from 'rxjs';

import {IConnectionSource} from '../connection';

import {IServerGenericParams} from './server';

export type QueuedChangeProcessor = (clock: number) => Promise<void>;

export interface BroadcastChangeResult {
  id: ChangePacketId;
  creations: ISyncable[];
  updates: ChangePlantProcessingResultUpdateItem[];
  removals: SyncableRef[];
}

export interface IServerAdapter<TGenericParams extends IServerGenericParams> {
  connectionSource$: Observable<IConnectionSource>;

  broadcast$: Observable<BroadcastChangeResult>;

  subscribe(group: string): Promise<void>;
  unsubscribe(group: string): Promise<void>;

  broadcast(group: string, data: BroadcastChangeResult): Promise<void>;

  queueChange(group: string, processor: QueuedChangeProcessor): Promise<void>;

  loadSyncablesByRefs(
    group: string,
    refs: SyncableRef<TGenericParams['syncableObject']>[],
  ): Promise<TGenericParams['syncableObject']['syncable'][]>;

  saveSyncables(
    group: string,
    createdSyncables: TGenericParams['syncableObject']['syncable'][],
    updatedSyncables: TGenericParams['syncableObject']['syncable'][],
    removedSyncableRefs: SyncableRef<TGenericParams['syncableObject']>[],
  ): Promise<void>;

  handleNotifications(
    group: string,
    notifications: TGenericParams['notification'][],
    id: ChangePacketId,
  ): Promise<void>;
}
