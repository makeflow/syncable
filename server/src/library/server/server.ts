import {IRPCDefinition, RPCFunctionDict} from '@syncable/core';

import {Connection, IConnectionSource} from '../connection';

import {IServerAdapter} from './server-adapter';

export class Server<TRPCDefinition extends IRPCDefinition> {
  constructor(
    private serverAdapter: IServerAdapter,
    private rpcFunctionDict: RPCFunctionDict<TRPCDefinition>,
  ) {
    serverAdapter.connectionSource$.subscribe(this.onConnectionSource);
  }

  private onConnectionSource = (source: IConnectionSource): void => {
    let connection = new Connection(source, this.rpcFunctionDict);
  };
}
