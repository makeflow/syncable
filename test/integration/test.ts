import {BroadcastChangeResult, Connection} from '@syncable/server';
import _ from 'lodash';
import Lolex, {Clock} from 'lolex';
import {Subject} from 'rxjs';

import {
  ServerGenericParams,
  TaskId,
  UserId,
  createClientConnectionPair,
  createServer,
} from './@setup';

let lolexClock: Clock;

beforeEach(() => {
  lolexClock = Lolex.install({
    now: 1500000000000,
    shouldAdvanceTime: true,
  });
});

afterEach(() => {
  lolexClock.uninstall();
});

let connection$ = new Subject<Connection<ServerGenericParams>>();
let broadcastSource$ = new Subject<BroadcastChangeResult>();

let server = createServer(connection$, broadcastSource$);

test('should initialize client with syncables and context data', async () => {
  let [client, connection, close] = createClientConnectionPair(
    server,
    'group-1',
    'user-1' as UserId,
  );

  connection$.next(connection);

  await client.ready;

  expect(_.cloneDeep(client.container.getSyncables())).toMatchSnapshot();
  expect(client.context.object.id).toBe('user-1');

  close();
});

test('should query tasks', async () => {
  let [client, connection, close] = createClientConnectionPair(
    server,
    'group-1',
    'user-1' as UserId,
  );

  connection$.next(connection);

  await client.query({
    task: {},
  });

  expect(_.cloneDeep(client.container.getSyncables('task'))).toMatchSnapshot();

  close();
});

test('should request tasks', async () => {
  let [client, connection, close] = createClientConnectionPair(
    server,
    'group-1',
    'user-1' as UserId,
  );

  connection$.next(connection);

  let tasks = await client.requestObjects([
    {type: 'task', id: 'task-1' as TaskId},
    {type: 'task', id: 'task-2' as TaskId},
  ]);

  expect(
    _.cloneDeep(tasks.map(task => task && task.syncable)),
  ).toMatchSnapshot();

  close();
});

test('should request single task', async () => {
  let [client, connection, close] = createClientConnectionPair(
    server,
    'group-1',
    'user-1' as UserId,
  );

  connection$.next(connection);

  let task = await client.requestObject({type: 'task', id: 'task-1' as TaskId});

  expect(_.cloneDeep(task && task.syncable)).toMatchSnapshot();

  close();
});

test('should update task brief', async () => {
  let [client, connection, close] = createClientConnectionPair(
    server,
    'group-1',
    'user-1' as UserId,
  );

  connection$.next(connection);

  await client.query({
    task: {},
  });

  let task = client.requireObject({type: 'task', id: 'task-2' as TaskId});

  await client.updateAndConfirm({
    type: 'task:update-task-brief',
    refs: {
      task: task.ref,
    },
    options: {
      brief: 'New brief for task 2',
    },
  });

  expect(_.cloneDeep(task && task.syncable)).toMatchSnapshot();

  close();
});
