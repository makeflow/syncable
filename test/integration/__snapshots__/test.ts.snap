// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should initialize client with syncables and context data 1`] = `
Array [
  Object {
    "_clock": 0,
    "_createdAt": 0,
    "_id": "user-1",
    "_type": "user",
    "_updatedAt": 0,
    "group": "group-1",
  },
]
`;

exports[`should query tasks 1`] = `
Array [
  Object {
    "_clock": 0,
    "_createdAt": 0,
    "_id": "task-1",
    "_type": "task",
    "_updatedAt": 0,
    "brief": "This is task 1",
    "group": "group-1",
  },
  Object {
    "_clock": 0,
    "_createdAt": 0,
    "_id": "task-2",
    "_type": "task",
    "_updatedAt": 0,
    "brief": "This is task 2",
    "group": "group-1",
  },
]
`;

exports[`should request single task 1`] = `
Object {
  "_clock": 0,
  "_createdAt": 0,
  "_id": "task-1",
  "_type": "task",
  "_updatedAt": 0,
  "brief": "This is task 1",
  "group": "group-1",
}
`;

exports[`should request tasks 1`] = `
Array [
  Object {
    "_clock": 0,
    "_createdAt": 0,
    "_id": "task-1",
    "_type": "task",
    "_updatedAt": 0,
    "brief": "This is task 1",
    "group": "group-1",
  },
  Object {
    "_clock": 0,
    "_createdAt": 0,
    "_id": "task-2",
    "_type": "task",
    "_updatedAt": 0,
    "brief": "This is task 2",
    "group": "group-1",
  },
]
`;

exports[`should update task brief: confirmed-update 1`] = `
Object {
  "_clock": 1,
  "_createdAt": 0,
  "_id": "task-2",
  "_type": "task",
  "_updatedAt": 1500000000240,
  "brief": "New brief for task 2",
  "group": "group-1",
}
`;

exports[`should update task brief: local-update 1`] = `
Object {
  "_clock": 0,
  "_createdAt": 0,
  "_id": "task-2",
  "_type": "task",
  "_updatedAt": 1500000000160,
  "brief": "New brief for task 2",
  "group": "group-1",
}
`;
